server/eclipse-project/src/main/java/us/freeandfair/corla/json/CVRContestInfoJsonAdapter.java

```diff
@@ -136,52 +149,101 @@ private Contest contestSanityCheck(final Long the_id,
-    
-    final Contest contest = contestSanityCheck(contest_id, choices);
-    
-    if (error || contest == null) {
-      throw new JsonSyntaxException("invalid data detected in CVR contest info");
+
+    // check the sanity of these choices for this contest. This means checking whether they can be
+    // parsed properly at all and, if so, whether the choices (candidate names or yes/no options)
+    // are the ones expected for this contest. For plurality, we expect plain choices that exactly
+    // match the valid choices; for IRV we expect something of the form "name(rank)" where "name"
+    // is a valid choice.
+    final Contest currentContest = Persistence.getByID(contest_id, Contest.class);
+
+    // For IRV contests, first check whether the choices can be parsed as a list of IRV
+    // preferences. Duplicates, overvotes and skipped ranks are OK, but strings that can't be
+    // parsed as name(rank) will throw an exception.
+    // Then conduct the sanity check (which checks whether the candidate names are the expected
+    // names for this contest) on just the names (with preferences removed).
+    //
+    // Note: a more succinct, but actually incorrect, way to do this would have been to get the
+    // IRV vote valid intent as
+    // List<String> validInOrder = new IRVChoices(choices).getValidIntentAsOrderedList();
+    // and then call contestSanityCheck on that. However, this would mean that invalid candidate
+    // selections that had been omitted as part of IRV valid interpretation would not be flagged.
+    // For example, a vote like [valid_candidate(1),invalid_candidate(3)] would first have the
+    // 3rd preference omitted (because preference 2 was skipped), and then have only
+    // [valid_candidate(1)] sanity-checked. The below implementation deliberately does this in the
+    // opposite order, first getting _all_ the mentioned candidates (as choicesForSanityChecking),
+    // sanity checking them, and then doing valid IRV interpretation afterwards, producing
+    // interpretedChoices, which is returned in the CVRContestInfo.
+
+    List<String> choicesForSanityChecking;
+    List<String> interpretedChoices;
+    try {
+      // For IRV, the choices for sanity checking (which includes all mentioned candidates) are
+      // different from the valid interpreted list (which omits everything after skipped or repeated
+      // prefernces).
+      if (currentContest.description().equalsIgnoreCase(ContestType.IRV.toString())) {
+        IRVChoices parsedChoices = new IRVChoices(choices);
+        choicesForSanityChecking = parsedChoices.getCandidateNames();
+        interpretedChoices = parsedChoices.GetValidIntentAsOrderedList();
+        // For plurality, just do the sanity check directly on the choices.
+      } else {
+        choicesForSanityChecking = choices;
+        interpretedChoices = choices;
+      }
+
+      Contest contest = contestSanityCheck(contest_id, choicesForSanityChecking);
+
```

Comment:
Log thrown exception; make format of message consistent with others.

Reply:
Done (actually this unlogged exception-throwing was already there, but now I guess we've made it better).